File: wpimath/src/test/native/cpp/kinematics/DifferentialDriveKinematicsTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <numbers>

#include <gtest/gtest.h>

#include "frc/kinematics/ChassisSpeeds.h"
#include "frc/kinematics/DifferentialDriveKinematics.h"
#include "frc/units.h"

using namespace frc;

static constexpr double kEpsilon = 1E-9;

TEST(DifferentialDriveKinematicsTest, InverseKinematicsFromZero) {
  const DifferentialDriveKinematics kinematics{2 * 0.381 * mp::m};
  const ChassisSpeeds chassisSpeeds;
  const auto wheelSpeeds = kinematics.ToWheelSpeeds(chassisSpeeds);

  EXPECT_NEAR(mp::value(wheelSpeeds.left), 0, kEpsilon);
  EXPECT_NEAR(mp::value(wheelSpeeds.right), 0, kEpsilon);
}

TEST(DifferentialDriveKinematicsTest, ForwardKinematicsFromZero) {
  const DifferentialDriveKinematics kinematics{2 * 0.381 * mp::m};
  const DifferentialDriveWheelSpeeds wheelSpeeds;
  const auto chassisSpeeds = kinematics.ToChassisSpeeds(wheelSpeeds);

  EXPECT_NEAR(mp::value(chassisSpeeds.vx), 0, kEpsilon);
  EXPECT_NEAR(mp::value(chassisSpeeds.vy), 0, kEpsilon);
  EXPECT_NEAR(mp::value(chassisSpeeds.omega), 0, kEpsilon);
}

TEST(DifferentialDriveKinematicsTest, InverseKinematicsForStraightLine) {
  const DifferentialDriveKinematics kinematics{2 * 0.381 * mp::m};
  const ChassisSpeeds chassisSpeeds{3.0 * mp::m / mp::s, 0.0 * mp::m / mp::s,
                                    0.0 * mp::rad / mp::s};
  const auto wheelSpeeds = kinematics.ToWheelSpeeds(chassisSpeeds);

  EXPECT_NEAR(mp::value(wheelSpeeds.left), 3, kEpsilon);
  EXPECT_NEAR(mp::value(wheelSpeeds.right), 3, kEpsilon);
}

TEST(DifferentialDriveKinematicsTest, ForwardKinematicsForStraightLine) {
  const DifferentialDriveKinematics kinematics{2 * 0.381 * mp::m};
  const DifferentialDriveWheelSpeeds wheelSpeeds{3.0 * mp::m / mp::s,
                                                 3.0 * mp::m / mp::s};
  const auto chassisSpeeds = kinematics.ToChassisSpeeds(wheelSpeeds);

  EXPECT_NEAR(mp::value(chassisSpeeds.vx), 3, kEpsilon);
  EXPECT_NEAR(mp::value(chassisSpeeds.vy), 0, kEpsilon);
  EXPECT_NEAR(mp::value(chassisSpeeds.omega), 0, kEpsilon);
}

TEST(DifferentialDriveKinematicsTest, InverseKinematicsForRotateInPlace) {
  const DifferentialDriveKinematics kinematics{2 * 0.381 * mp::m};
  const ChassisSpeeds chassisSpeeds{0.0 * mp::m / mp::s, 0.0 * mp::m / mp::s,
                                    std::numbers::pi * mp::rad / mp::s};
  const auto wheelSpeeds = kinematics.ToWheelSpeeds(chassisSpeeds);

  EXPECT_NEAR(mp::value(wheelSpeeds.left), -0.381 * std::numbers::pi, kEpsilon);
  EXPECT_NEAR(mp::value(wheelSpeeds.right), +0.381 * std::numbers::pi,
              kEpsilon);
}

TEST(DifferentialDriveKinematicsTest, ForwardKinematicsForRotateInPlace) {
  const DifferentialDriveKinematics kinematics{2 * 0.381 * mp::m};
  const DifferentialDriveWheelSpeeds wheelSpeeds{
      +0.381 * std::numbers::pi * mp::m / mp::s,
      -0.381 * std::numbers::pi * mp::m / mp::s};
  const auto chassisSpeeds = kinematics.ToChassisSpeeds(wheelSpeeds);

  EXPECT_NEAR(mp::value(chassisSpeeds.vx), 0, kEpsilon);
  EXPECT_NEAR(mp::value(chassisSpeeds.vy), 0, kEpsilon);
  EXPECT_NEAR(mp::value(chassisSpeeds.omega), -std::numbers::pi, kEpsilon);
}

File: wpimath/src/test/java/org/wpilib/math/kinematics/DifferentialDriveKinematicsTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.math.kinematics;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

class DifferentialDriveKinematicsTest {
  private static final double EPSILON = 1E-9;
  private final DifferentialDriveKinematics m_kinematics =
      new DifferentialDriveKinematics(0.381 * 2);

  @Test
  void testInverseKinematicsForZeros() {
    var chassisSpeeds = new ChassisSpeeds();
    var wheelSpeeds = m_kinematics.toWheelSpeeds(chassisSpeeds);

    assertAll(
        () -> assertEquals(0.0, wheelSpeeds.left, EPSILON),
        () -> assertEquals(0.0, wheelSpeeds.right, EPSILON));
  }

  @Test
  void testForwardKinematicsForZeros() {
    var wheelSpeeds = new DifferentialDriveWheelSpeeds();
    var chassisSpeeds = m_kinematics.toChassisSpeeds(wheelSpeeds);

    assertAll(
        () -> assertEquals(0.0, chassisSpeeds.vx, EPSILON),
        () -> assertEquals(0.0, chassisSpeeds.vy, EPSILON),
        () -> assertEquals(0.0, chassisSpeeds.omega, EPSILON));
  }

  @Test
  void testInverseKinematicsForStraightLine() {
    var chassisSpeeds = new ChassisSpeeds(3, 0, 0);
    var wheelSpeeds = m_kinematics.toWheelSpeeds(chassisSpeeds);

    assertAll(
        () -> assertEquals(3.0, wheelSpeeds.left, EPSILON),
        () -> assertEquals(3.0, wheelSpeeds.right, EPSILON));
  }

  @Test
  void testForwardKinematicsForStraightLine() {
    var wheelSpeeds = new DifferentialDriveWheelSpeeds(3, 3);
    var chassisSpeeds = m_kinematics.toChassisSpeeds(wheelSpeeds);

    assertAll(
        () -> assertEquals(3.0, chassisSpeeds.vx, EPSILON),
        () -> assertEquals(0.0, chassisSpeeds.vy, EPSILON),
        () -> assertEquals(0.0, chassisSpeeds.omega, EPSILON));
  }

  @Test
  void testInverseKinematicsForRotateInPlace() {
    var chassisSpeeds = new ChassisSpeeds(0, 0, Math.PI);
    var wheelSpeeds = m_kinematics.toWheelSpeeds(chassisSpeeds);

    assertAll(
        () -> assertEquals(-0.381 * Math.PI, wheelSpeeds.left, EPSILON),
        () -> assertEquals(+0.381 * Math.PI, wheelSpeeds.right, EPSILON));
  }

  @Test
  void testForwardKinematicsForRotateInPlace() {
    var wheelSpeeds = new DifferentialDriveWheelSpeeds(+0.381 * Math.PI, -0.381 * Math.PI);
    var chassisSpeeds = m_kinematics.toChassisSpeeds(wheelSpeeds);

    assertAll(
        () -> assertEquals(0.0, chassisSpeeds.vx, EPSILON),
        () -> assertEquals(0.0, chassisSpeeds.vy, EPSILON),
        () -> assertEquals(-Math.PI, chassisSpeeds.omega, EPSILON));
  }

  @Test
  void testInverseAccelerationsForZeros() {
    var chassisAccelerations = new ChassisAccelerations();
    var wheelAccelerations = m_kinematics.toWheelAccelerations(chassisAccelerations);

    assertAll(
        () -> assertEquals(0.0, wheelAccelerations.left, EPSILON),
        () -> assertEquals(0.0, wheelAccelerations.right, EPSILON));
  }

  @Test
  void testForwardAccelerationsForZeros() {
    var wheelAccelerations = new DifferentialDriveWheelAccelerations();
    var chassisAccelerations = m_kinematics.toChassisAccelerations(wheelAccelerations);

    assertAll(
        () -> assertEquals(0.0, chassisAccelerations.ax, EPSILON),
        () -> assertEquals(0.0, chassisAccelerations.ay, EPSILON),
        () -> assertEquals(0.0, chassisAccelerations.alpha, EPSILON));
  }

  @Test
  void testInverseAccelerationsForStraightLine() {
    var chassisAccelerations = new ChassisAccelerations(3, 0, 0);
    var wheelAccelerations = m_kinematics.toWheelAccelerations(chassisAccelerations);

    assertAll(
        () -> assertEquals(3.0, wheelAccelerations.left, EPSILON),
        () -> assertEquals(3.0, wheelAccelerations.right, EPSILON));
  }

  @Test
  void testForwardAccelerationsForStraightLine() {
    var wheelAccelerations = new DifferentialDriveWheelAccelerations(3, 3);
    var chassisAccelerations = m_kinematics.toChassisAccelerations(wheelAccelerations);

    assertAll(
        () -> assertEquals(3.0, chassisAccelerations.ax, EPSILON),
        () -> assertEquals(0.0, chassisAccelerations.ay, EPSILON),
        () -> assertEquals(0.0, chassisAccelerations.alpha, EPSILON));
  }

  @Test
  void testInverseAccelerationsForRotateInPlace() {
    var chassisAccelerations = new ChassisAccelerations(0, 0, Math.PI);
    var wheelAccelerations = m_kinematics.toWheelAccelerations(chassisAccelerations);

    assertAll(
        () -> assertEquals(-0.381 * Math.PI, wheelAccelerations.left, EPSILON),
        () -> assertEquals(+0.381 * Math.PI, wheelAccelerations.right, EPSILON));
  }

  @Test
  void testForwardAccelerationsForRotateInPlace() {
    var wheelAccelerations =
        new DifferentialDriveWheelAccelerations(+0.381 * Math.PI, -0.381 * Math.PI);
    var chassisAccelerations = m_kinematics.toChassisAccelerations(wheelAccelerations);

    assertAll(
        () -> assertEquals(0.0, chassisAccelerations.ax, EPSILON),
        () -> assertEquals(0.0, chassisAccelerations.ay, EPSILON),
        () -> assertEquals(-Math.PI, chassisAccelerations.alpha, EPSILON));
  }
}
